core/domain/role_services.py

```diff
@@ -131,3 +168,41 @@ def get_all_actions(role):
         role_actions.extend(get_all_actions(parent_role))
 
     return list(set(role_actions))
+
+
+def get_role_graph_data():
+    """Returns dict for displaying roles graph.
+
+    Returns:
+        dict. A dict containing data in following format:
+        {
+            links: list(dict(str:str)). List of dicts defing each edge in
+                following format:
+                    {
+                        source: Role Id from which edge is going out.
+                        target: Role Id to which edge is incoming.
+                    }
+            nodes: dict(str:str). Mapping of role Id to its human readable
+                format.
+        }
+    """
+    role_graph = {}
+    role_graph['links'] = []
+    role_graph['nodes'] = {}
+    for role in PARENT_ROLES:
+        role_graph['nodes'][role] = HUMAN_READABLE_ROLES[role]
+        for parent in PARENT_ROLES[role]:
+            role_graph['links'].append({'source': parent, 'target': role})
+    return role_graph
+
+
+def store_role_query(user_id, intent, role=None, username=None):
+    """Stores the query to role structure in RoleQueryAuditModel."""
+    model_id = (
+        str(user_id) + '.' + str(int(math.floor(time.time()))) + '.' +
+        str(intent) + '.' + str(random.randint(0, 1000))
+    )
+
+    audit_models.RoleQueryAuditModel(
+        model_id=model_id, user_id=user_id, intent=intent,
```

Comment:
model_id --> id (this is a standard field in all the models, otherwise an id gets autogenerated by App Engine)